### PR TITLE
extensible variable semantics

### PR DESCRIPTION
**Pull request by exaexa**
Not really working yet, but this should clear the current mess in variables vs. reactions vs. fluxes vs. enzymes vs. mass groups vs. whatnot.

### CODE REVIEW

**Review by stelmo — APPROVED** [review]
lg2m

**Comment by exaexa** [comment]
@stelmo there are some TODOs before I can merge this, but the API for variables is final. I'll do the constraint APIs in a separate PR.

File: src/analysis/flux_variability_analysis.jl [hunk 1]
```diff
@@ -115,7 +115,7 @@ function flux_variability_analysis(
 
     flux_variability_analysis(
         model,
-        reaction_variables(model)[:, flux_indexes],
+        reaction_variables_matrix(model)[:, flux_indexes],
```
Comment (stelmo) [inline src/analysis/flux_variability_analysis.jl:118]: at the price of making things a lot more verbose, how about `reaction_variables_mapping_matrix`?

File: src/types/models/SBMLModel.jl [hunk 2]
```diff
@@ -45,6 +45,8 @@ Get variables (aka reactions) from a [`SBMLModel`](@ref).
 """
 Accessors.variables(model::SBMLModel)::Vector{String} = model.reaction_ids
 
+Accessors.Internal.@all_variables_are_reactions SBMLModel
```
Comment (stelmo) [inline src/types/models/SBMLModel.jl:48]: smart

File: src/types/accessors/bits/semantics.jl [hunk 3]
```diff
@@ -0,0 +1,148 @@
+function make_mapping_mtx(rows, cols, col_row_val)
```
Comment (stelmo) [inline src/types/accessors/bits/semantics.jl:1]: This needs a docstring and an example to explain how to use it

File: src/types/accessors/bits/semantics.jl [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    # TODO docs
+    quote
+        $Accessors.reactions(model::$m) = $Accessors.variables(model)
+        $Accessors.n_reactions(model::$m) = $Accessors.n_variables(model)
+        $Accessors.reaction_variables(model::$m) =
+            Dict(var => Dict(var => 1.0) for var in variables(model))
+        $Accessors.reaction_variables_matrix(model::$m) =
+            $SparseArrays.spdiagm(fill(1.0, $Accessors.n_variables(model)))
+    end
+end
```
Comment (stelmo) [inline src/types/accessors/bits/semantics.jl:148]: nice

Comment (stelmo) [inline src/types/accessors/bits/semantics.jl:148]: All of this needs a "how to for dummies" - I will test out if it is clear enough when I do the enzyme models later today

**Review by stelmo — APPROVED** [review]
Really good work

File: src/analysis/flux_variability_analysis.jl (hunk 1 above)
Comment (exaexa) [reply src/analysis/flux_variability_analysis.jl]: "mapping" already sounds like "bipartite"... I'll keep it short now

File: src/types/accessors/bits/semantics.jl (hunk 4 above)
Comment (exaexa) [reply src/types/accessors/bits/semantics.jl]: I added some docstrings; dummies may find the basic examples with grep. There will also be docs.

**Pull request merged**